### DISCUSSION TITLE
DGAL, Ortho4XP, macOX 11 (Big Sur) and Python 3.9

### DISCUSSION THREAD

**Discussion by OldSubSailor** [Homebrew/discussions#367] [Everyday usage] [answered]
All of these thing are somewhat interrelated. Until I upgraded to Big Sur, Ortho4XP and GDAL was working fine.  I know next to nothing about python, but I eventually did find these lines of code in Ortho4XP:

```
import gdal
has_gdal=True
except:
has_gdal=False
```

which leads me to think that GDAL that we get from Homebrew is supposed to be a python import module and was working as expected. 
Then came the upgrade to Big Sur and python 3.9 and things went south.  Ortho4XP only uses GDAL when we create elevation tiles from USGS NED data.  So when I try that I ultimately end up with this:

```
elif not has_gdal:
UI.lvprint(1," WARNING: unsupported raster (install Gdal):", file_name, "-> replaced with zero altitude.")
nxdem=nydem=base_if_error
```
```
error message telling me to install GDAL which I did install via home brew, and shows up in a brew listing:
% brew info GDAL
gdal: stable 3.2.0 (bottled), HEAD
```

I also tried calling the module directly from python:
```
Python 3.9.0 (v3.9.0:9cf6752276, Oct  5 2020, 11:29:23) 
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import gdal
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'gdal'
```

I updated homebrew and GDAL but still GDAL fails. 
I would appreciate your help. 

Update, I also changed to Python 3.8 but no luck.  However, I did see that a fresh homebrew install of GDAL, adds some python 3.9 to things. But calling python3 still brings up 3.8 at this time.

**Comment by OldSubSailor** [comment]
No answers; HOWEVER; A FIX!!  Today, I upgraded and updated Homebrew and installed packages and like magic it works again. Why, I do not know, except evidently everything is  flanged up once more.

**Reply by OldSubSailor** [reply]
Well, perhaps not a fix after all.  I updated about 5 tiles then on the 6th the old error message was back again. I did not change a thing. 
The Ortho4XP error message reads:
Loading elevation data and smoothing it over airports.
   WARNING: unsupported raster (install Gdal): ./Elevation_data/+30-100/N31W095_NED13.img -> replaced with zero altitude.
    * Min altitude: 0.0 , Max altitude: 0.0 , Mean: 0.0
 
Gdal is installed. 

 brew info gdal
gdal: stable 3.3.1 (bottled), HEAD
Geospatial Data Abstraction Library
https://www.gdal.org/
Conflicts with:
  avce00 (because both install a cpl_conv.h header)
  cpl (because both install cpl_error.h)
/usr/local/Cellar/gdal/3.3.1_2 (356 files, 58.9MB) *
  Poured from bottle on 2021-07-29 at 12:05:06
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/gdal.rb
License: MIT
==> Dependencies
Build: pkg-config ✔
Required: cfitsio ✔, epsilon ✔, expat ✔, freexl ✔, geos ✔, giflib ✔, hdf5 ✔, jpeg ✔, json-c ✔, libdap ✔, libgeotiff ✔, libpng ✔, libpq ✔, libspatialite ✔, libtiff ✔, libxml2 ✔, netcdf ✔, numpy ✔, openjpeg ✔, pcre ✔, poppler-qt5 ✔, proj@7 ✔, python@3.9 ✔, sqlite ✔, unixodbc ✔, webp ✔, xerces-c ✔, xz ✔, zstd ✔
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 20,044 (30 days), 56,294 (90 days), 233,264 (365 days)
install-on-request: 16,170 (30 days), 44,893 (90 days), 179,450 (365 days)
build-error: 0 (30 days)
% 

and the import call from python once again does not work.

**Comment by OldSubSailor** [comment]
Warning: proj@7 has been deprecated because it is not supported upstream!
Evidently required by Gdal, but this warning is generated when it is installed. Could it be causing problems. I have no idea what "not supported upstream" means.

**Reply by gromgit** [reply]
> I have no idea what "not supported upstream" means.

PROJ has neither a proven history nor stated intent to fix earlier releases. Indeed, it seems to adopt the "EOL everything except current version" mindset that is common with open-source projects.

In fact, by failing at least that [criterion for formula versioning](https://docs.brew.sh/Versions), `proj@7` shouldn't exist. It was only created because 9 other formulae couldn't (and perhaps still can't) build with PROJ 8: https://github.com/Homebrew/homebrew-core/pull/77693.

**Comment by OldSubSailor** [comment]
Thank you for your reply,
Does that mean that Gdal will be corrected at some time to change dependancies?

**Reply by gromgit** [reply]
Only if _all_ its dependencies stop relying on `proj@7`.

**Reply by OldSubSailor** [reply]
Drat! Is there some Wizard who fixes that kind of stuff? Do I need to put out a bow of milk? Heck, I'm 74 years old, I aint got much longer to wait.
When I say OLD SUB SAILOR, I mean it, the first boat I rode was the 571 boat.

**Reply by SMillerDev** [reply]
You can create a pull request to the GDAL project to fix it. It's an open source project, so it's up to the public to fix issues.

**Comment by OldSubSailor** [comment]
Wow, this sounds like a potential mess? 
Unfortunately for a brief moment, it things were working with Ortho4-Xplane again and then stopped. 
Although Gdal shows up in the brew listing, I get the module not found error if I try the import command from python once more. 

>>> import gdal
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'gdal'
>>> 


alas.
All way above my old head.

**Comment by OldSubSailor** [comment]
For some reason, when I try doing that I mess it up and I fail.

**Comment by OldSubSailor** [comment]
> PULL REQUEST SUBMITTED TO GDAL
> During the downloading process, this warning appears.
> Warning: proj@7 has been deprecated because it is not supported upstream!
> proj@7 is listed as a dependency, however, I know nothing about these matters other than searches indicate that Gdal is installed on my iMac (Big Sur)
> % brew info gdal
> gdal: stable 3.3.1 (bottled), HEAD
> Geospatial Data Abstraction Library
> However, it appears not to be working, and a direct call from python says that the the module is missing.
> I have reported it to Homebrew, and they say to submit a pull request to gdal.
> I feel like I am caught in a Catch 22.
> I don't think that this is a support request, but a concern developer intervention.


This is the response that I got from Gdal.  NOTE that I said that I had REPORTED it to HOMEBREW.
This is not appropriate as a pull request, and actually not even as a bug report here. This is to report to the homebrew maintainers for GDAL


#4159 <https://github.com/OSGeo/gdal/pull/4159>

**Reply by OldSubSailor** [reply]
So are there homebrew maintainers of GDAL?

**Reply by SMillerDev** [reply]
I'm not exactly sure what that pull request was supposed to do, but the pull request that would be appropriate to upstream to GDAL would be one that adds support for proj 8 to GDAL.

**Reply by OldSubSailor** [reply]
I don't know why proj@7 was blanked out in the warning that I posted above.

**Comment by OldSubSailor** [comment]
The question is pull request would to to whom and where?

**Comment by OldSubSailor** [comment]
I am trying to create a pull request, but am stumbling about in GitHub Homebrew trying to find where. Evidently I can’t create a branch in which to generate a new pull request, and am unsuccessful getting a comparison match on branches to write a pull request at this time.  

I thought that one clicked on the main branch drop down to make a new branch to try and make a pull request.

**Comment by jef-n** [comment, marked as answer]
Try ```import osgeo.gdal``` instead of ```import gdal```

**Reply by OldSubSailor** [reply]
thank you for your reply. 
I still get the same module not found. The brew listing only shows gdal.

**Reply by OldSubSailor** [reply]
the import osgeo.gdal is the answer after all.

**Reply by OldSubSailor** [reply]
But as of this summer it only works for USGS 1".
